torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -406,30 +406,51 @@ def sync(self) -> Optional[bool]:
         return has_set
 
     def _sanitize(self) -> None:
+        state = self._state
+
         expire_time = datetime.utcnow() - (
             self._settings.keep_alive_interval * self._settings.keep_alive_max_attempt
         )
 
         # Filter out the dead nodes.
         self._dead_nodes = [
             node
-            for node, last_heartbeat in self._state.last_heartbeats.items()
+            for node, last_heartbeat in state.last_heartbeats.items()
             if last_heartbeat < expire_time
         ]
 
+        participant_removed = False
+
         for dead_node in self._dead_nodes:
-            del self._state.last_heartbeats[dead_node]
+            del state.last_heartbeats[dead_node]
 
             try:
-                del self._state.participants[dead_node]
+                del state.participants[dead_node]
+
+                participant_removed = True
             except KeyError:
                 pass
 
             try:
-                self._state.wait_list.remove(dead_node)
+                state.wait_list.remove(dead_node)
             except KeyError:
                 pass
 
+        if not participant_removed:
+            return
+
+        # If a participant is removed from the rendezvous, we have to make sure
+        # that the state of the rendezvous reflects it.
+        if state.complete:
```

Comment:
I agree. First I was a bit hesitant to merge them since that would likely make reading the code slightly harder. Now seeing that you are having the same thought, I decided to give it a shot. Check out the second rev.